soroban-env-common/src/env_val.rs

```diff
@@ -12,22 +13,27 @@ use super::{
     Env,
 };
 
-pub trait IntoVal<E: Env, V>: Sized {
-    fn into_val(self, env: &E) -> V;
+pub trait TryIntoVal<E: Env, V> {
+    type Error;
+    fn try_into_val(&self, env: &E) -> Result<V, Self::Error>;
 }
 
-pub trait TryIntoVal<E: Env, V>: Sized {
+pub trait TryFromVal<E: Env, V: ?Sized>: Sized {
     type Error;
-    fn try_into_val(self, env: &E) -> Result<V, Self::Error>;
+    fn try_from_val(env: &E, v: impl Borrow<V>) -> Result<Self, Self::Error>;
 }
 
-pub trait FromVal<E: Env, V>: Sized {
-    fn from_val(env: &E, v: V) -> Self;
-}
+// Blanket-impl uses of TryIntoVal to TryFromVal, so that we
+// only ever have to impl TryFromVal.
+impl<E: Env, T, U> TryIntoVal<E, T> for U
```

Comment:
We used to have a blanket impl like this, and it had to be removed at some point for another reason. I think it related to UDTs, but I can't remember exactly. If we can have it again sounds good to me, but I wonder if somewhere downstream we'll see an issue with this. It's also possible the blanket impl was slightly different and that this doesn't face the same problems.